### PR TITLE
Implement the new bounding box API

### PR DESCRIPTION
**Pull request by LukeWood**

### CODE REVIEW

File: keras_cv/bounding_box/__init__.py
Comment (kartik4949) [inline keras_cv/bounding_box/__init__.py:28]: extended_corners format, 
one of the KPI needed it guess.

Comment (LukeWood) [reply keras_cv/bounding_box/__init__.py:28]: This can be done in a follow up 👍

Comment (kartik4949) [reply keras_cv/bounding_box/__init__.py:28]: yup, make sense, can be done in next pr :)

**Review by divyashreepathihalli — APPROVED** [review]

File: keras_cv/bounding_box/converters_text.py
Comment (quantumalaviya) [inline keras_cv/bounding_box/converters_text.py:1]: Shouldn't this file rather be named `converters_test.py` instead of `converters_text.py`?

Comment (LukeWood) [reply keras_cv/bounding_box/converters_text.py:1]: Wow good catch, thanks

**Comment by LukeWood** [comment]
gentle ping @qlzh727 @divyashreepathihalli and @fchollet - lets prioritize this PR if we can.

File: keras_cv/metrics/coco/mean_average_precision.py
Comment (qlzh727) [inline keras_cv/metrics/coco/mean_average_precision.py:162]: Is the casting still needed here?

File: keras_cv/bounding_box/converters_text.py
Comment (qlzh727) [inline keras_cv/bounding_box/converters_text.py:22]: Let's add a test for int dtype here.

**Review by qlzh727 — CHANGES_REQUESTED** [review]
Sorry for the late reply.

File: keras_cv/bounding_box/converters.py
Comment (fchollet) [reply keras_cv/bounding_box/converters.py:78]: Format the names as strings with quotes and backticks.

File: keras_cv/bounding_box/converters.py
Comment (fchollet) [inline keras_cv/bounding_box/converters.py:118]: Same

File: keras_cv/bounding_box/converters.py
Comment (fchollet) [inline keras_cv/bounding_box/converters.py:99]: That should be "source" not "original".

File: keras_cv/bounding_box/__init__.py
Comment (LukeWood) [reply keras_cv/bounding_box/__init__.py:27]: I think we have to standardize on some axes here.  To make use of our Loss functions, COCO metrics, or anything else like this we *have* to standardize on something or other.  I'd rather document them here, give users a constant to rely on, etc.

Users don't *have* to include these to use our bounding box conversion utility, as long as every axis after the `N` required to store the bounding box itself come first.  

In short, I think we need to offer these constants, but should write utilities with the expectation some users will not provide these.  Losses, metrics, layers, etc will need these though so I think we do need these.

File: keras_cv/bounding_box/converters.py
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py:74]: I think I like the idea of having a public namespace `bounding_box`. 

`bounding_box.transformat_format()` is a nice API name, but `transform_format()` is not.

I guess it depends on how we think users will handle their imports.  If they use the Keras style:
```
from keras_cv import bounding_box
```
or

```
from keras_cv import bounding_box as bounding_box_utils
```

then this `convert_format` or `transform` or `convert` is nice, otherwise having bounding_box in the name is a bit nicer.

Thoughts?

File: keras_cv/bounding_box/converters_text.py
Comment (LukeWood) [reply keras_cv/bounding_box/converters_text.py]: Maybe we should assert that the inputs are a float?

File: keras_cv/bounding_box/converters.py
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py:86]: yeah, it is optional.   Users can just pass `xywh` - it gets lower cased anyways.  I suggested this because I thought it may improve readability, but we can remove it.

File: keras_cv/bounding_box/converters.py
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py]: Done

File: keras_cv/bounding_box/converters.py
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py]: We can remove this if we think it hurts readability

File: keras_cv/metrics/coco/mean_average_precision.py
Comment (LukeWood) [reply keras_cv/metrics/coco/mean_average_precision.py:162]: Updated to do it internally, this is required for many conversions.

**Review by qlzh727 — APPROVED** [review]
LGTM. Just some minor comments.

File: keras_cv/bounding_box/converters.py
Comment (fchollet) [reply keras_cv/bounding_box/converters.py:31]: This will never be surfaced to users, right?

In general we should never raise a custom exception to users. But if we're sure that it will be caught internally and reraised as a regular exception this is fine.

File: keras_cv/bounding_box/converters.py
Comment (fchollet) [reply keras_cv/bounding_box/converters.py]: In Keras we used to have `keras.utils.image_utils` and so on, but we ended up moving everything to a flat `keras.utils` namespace. Overall, no strong opinion here, at least `keras_cv.bounding_box` is not nested. We can keep `keras_cv.bounding_box.convert_format`.

I'd definitely favor "convert" over "transform" though.

File: keras_cv/bounding_box/converters.py
Comment (tanzhenyu) [inline keras_cv/bounding_box/converters.py:163]: no special case for source == target?

File: keras_cv/bounding_box/converters_test.py
Comment (tanzhenyu) [reply keras_cv/bounding_box/converters_test.py:22]: +1 on ragged test

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!

File: keras_cv/metrics/coco/numerical_tests/recall_correctness_test.py
Comment (LukeWood) [inline keras_cv/metrics/coco/numerical_tests/recall_correctness_test.py:129]: favor convert terminology over transform terminology

File: keras_cv/bounding_box/converters.py
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py:31]: This will never be raised to users, it is an implementation detail only.

File: keras_cv/bounding_box/converters.py
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py]: Updated docstring.

File: keras_cv/bounding_box/converters.py
Comment (LukeWood) [reply keras_cv/bounding_box/converters.py:164]: Good catch, added

File: keras_cv/bounding_box/converters_test.py
Comment (LukeWood) [reply keras_cv/bounding_box/converters_test.py:22]: > Can we update the unit test to let images has different number of bbox within a batch (ragged)? or we have to ask user to pad the bbox for now?

Let's leave this as dense only for now to unblock the efforts blocked on this PR.

 I opened https://github.com/keras-team/keras-cv/issues/478 to track.

**Pull request merged**
